votes: propose a release notes language style guide (#639)
pingcap/community: votes/0639-propose-a-release-note-guide.md

- Compatibility change
- Improvement or Feature enhancement
- Bug fix

## Compatibility change

A compatibility change note means:

- Your PR adds, removes, or modifies one or more configuration items or system variables.
- Your PR modifies the default value of a configuration item or system variable.

For this type of note, you should clearly and adequately state the following aspects:

- The previous code behavior, configuration item, or default value.
- The new code behavior, configuration item, or default value since the new version.

Note that the object of the change should be user-perceivable. If the changed configuration item or system variable is not supposed to be exposed to users, do not include it in your release notes.

**Examples:**

| **Not recommended**                                          | **Clear in type** | **Adequate and clear in meaning** | **User perspective** | Recommended                                                  |
| ------------------------------------------------------------ | ----------------- | --------------------------------- | -------------------- | ------------------------------------------------------------ |

> | copr: cast invalid utf8 string to real bug fix               | ❌                 | ❌                                 | ❌                    | Before v4.0.16, when TiDB converts an illegal UTF-8 string to a Real type, an error is reported directly. Starting from v4.0.16, TiDB processes the conversion according to the legal UTF-8 prefix in the string |

Reviewer: I wonder if it is a good idea to mention version number in release note. Consider the PR (containing release note) can cherry-pick to different release branches, we will have to manually update the version number.

Author: Each release notes file is dedicated to a specific version, for example, v4.0.16. Even if this file is cherry-picked to other branches, it still only applies to that specific version (v4.0.16). For preparation, each release notes file is prepared separately, and the manual labor to change version number is almost negligible compared with the total effort required to prepare the entire release notes for each release.

Reviewer: I thought release notes are written in PR description and we have some tool to collect notes and generate full release note for each version.

Author: Yes, you are right. That's exactly how we are collecting release notes from the PR description part. The problem is that the collected raw material might not always be usable or deliverable. That's why we propose this release note guide. The aim is to help developers write more usable or deliverable notes.

Reviewer: The problem is that when we write release note in a PR. We cannot know what version will contain the patch and what branches the PR will be cherry-picked.

Author: I see. I'll remove the version number from the example. PTAL whether this modification is OK. Thanks~
```suggestion
| copr: cast invalid utf8 string to real bug fix               | ❌                 | ❌                                 | ❌                    | Previously, when TiDB converts an illegal UTF-8 string to a Real type, an error is reported directly. From now on, TiDB will process the conversion according to the legal UTF-8 prefix in the string. |
```

Reviewer 2: I think this release note guide seems to be more useful when it is applied to the process of writing a release note for a release by PMs, **instead of to be used in PRs**, for these reasons:

1. Multiple PRs may contribute to the same product behavior. For example, improve performance, improve performance, ... . It is not necessary to expand the details for each one.
2. Most of the developers will not get it right to be user perspective. Developers are not PMs.
3. A friendly release note contains much more than what developers care about. For example, previous behaviors, changed versions (mentioned by @disksing in this thread), etc. Developers is very likely to not help you do it as it is out of responsibility.

I mean, maybe our PMs should all read this, and strictly follow this :) Instead of devs. However it is always good to have some attempts. Keep going!

> | sink: fix kafka max message size inaccurate issue            | ❌                 | ❌                                 | ❌                    | Change the default value of Kafka Sink max-message-bytes to 1 MB to prevent TiCDC from sending too large messages to Kafka clusters |

Reviewer 3: this recommended stating does not include "The previous code behavior, configuration item, or default value.". I guess because the PR author can't infer the previous behaviour from given information?

Author: ```suggestion
| sink: fix kafka max message size inaccurate issue            | ❌                 | ❌                                 | ❌                    | Change the default value of Kafka Sink max-message-bytes from 512 MB to 1 MB to prevent TiCDC from sending too large messages to Kafka clusters |
```

| cdc/sink: adjust kafka initialization logic                  | ❌                 | ❌                                 | ❌                    | Change the default value of Kafka Sink partition-num to 3 so that TiCDC distributes messages across Kafka partitions more evenly |
| cmd: hide --sort-dir in changefeed command. (deprecated warning exists) | ✅                 | ❌                                 | ❌                    | Deprecate `--sort-dir` in the `cdc cli changefeed` command. Instead, users can set `--sort-dir` in the `cdc server` command. |

## Bug fix

A bug fix note means that your PR fixes an existing bug or issue. This type of notes start with "Fix" followed by "the issue/bug".

> Write your note clearly and adequately so that your target readers can get the main point of your bug fix. The bug or issue must be directly perceivable to the users, and you can refer to the associated GitHub issues.

Reviewer 2: Should we also recommend to highlight the bug trigger condition? It may be a good practice.

For example:

✓: Fix the issue that the computed result can be wrong when feature foo is enabled
✗: Fix the issue that the computed result can be wrong

Author: Added

**Examples:**

| **Not recommended**                                          | **Clear in type** | **Adequate and clear in meaning** | **User perspective** | Recommended                                                  |
| ------------------------------------------------------------ | ----------------- | --------------------------------- | -------------------- | ------------------------------------------------------------ |
| lock_resolver: avoid pessimistic transactions using resolveLocksForWrite | ❌                 | ✅                                 | ❌                    | Fix the issue that committing pessimistic transactions might cause write conflict |

> | retry when meeting stablish conn fails                       | ❌                 | ❌                                 | ❌                    | Fix the issue of unexpected results when TiFlash fails to establish MPP connections |

Reviewer 3: from the developer's view, I only solved one internal problem. But this problem may cause many problems at the user's view, I hope I can only list a few of them at my best and must include the one that reported in Github issue?

Author: It is a question of who decides what to include in release notes. Ideally, it's decided by product managers, but in fact, it's decided by each developer and the PR reviewers. If a developer thinks that his or her PR is worthy of a release note (users need to know), then he or she writes it in the PR description part, and vice versa. Maybe in the future and hopefully, our PM can also have a say in this respect.

| Fix the issue that greatest(datetime) union null returns empty string | ✅                 | ❌                                 | ✅                    | Fix the issue that the query result might be wrong when NULL is in the UNION subquery |

> | copr: make CM Sketch built with the same encoding as what TiDB assumes | ❌                 | ❌                                 | ❌                    | Fix the wrong statistics caused by the format inconsistency of CM Sketch between TiKV and TiDB |

Reviewer 3: I'm not sure if we can assume the user as a naive user or a sophisticated user who knows `CM Sketch`. So to what extend can we use technical terms?

Author: User perspective sometimes might be tricky. I think that the PM's role is needed here.

> | copr: make CM Sketch built with the same encoding as what TiDB assumes | ❌                 | ❌                                 | ❌                    | Fix the wrong statistics caused by the format inconsistency of CM Sketch between TiKV and TiDB |

Reviewer 2: From the user perspective we should describe when this problem will happen. It is what user really cares about. The CM Sketch is not a concept exposed to user so that it should not be mentioned.

```suggestion
| copr: make CM Sketch built with the same encoding as what TiDB assumes | ❌                 | ❌                                 | ❌                    | Fix wrong statistics when `tidb_analyze_version` is set to `1` |
```

Reviewer 2: I see that all bug fix notes are actually prefixed with "Fix the issue that ...". Is it possible to be simplified? For example, how about "Fix the analyzed statistics might be wrong when `tidb_analyze_version` is set to `1`"?

Author: I don't think "Fix the analyzed statistics might be" is grammatically correct. 

"Fix the issue" can have some variations, such as "Fix the issue of potential wrong analyzed statistics when ...".

## Improvement

An improvement note means that your PR improves stability or performance of the product, or enhances an existing feature. In addition to describing what your PR has changed, you should also mention how users can benefit from it.

This type of release note consists of two parts: what you have changed + the benefit of your change. This type of release notes often starts with "support", "increase", "improve", "optimize", etc.

**Examples:**

| **Not recommended**                                          | **Clear in type** | **Adequate and clear in meaning** | **User perspective** | Recommended                                                  |
| ------------------------------------------------------------ | ----------------- | --------------------------------- | -------------------- | ------------------------------------------------------------ |

> | Not use the stale read request's `start_ts` to update `max_ts` to avoid commit request keep retrying | ✅                 | ✅                                 | ✅                    | Avoid excessive commit request retrying by not using the Stale Read request's start_ts to update `max_ts` |

Reviewer 2: As a user I don't understand what is "commit request", "Stale Read request", "start_ts" and "max_ts".

Author: I think you have spotted the real problem. When we prepare release notes for a release, a PM's role is indeed expected. But sometimes, the reality is different from the ideal situation. We do not have enough PMs. Currently, we can only hope that each developer can learn to understand the users and that's also expected from the higher levels.

> | Not use the stale read request's `start_ts` to update `max_ts` to avoid commit request keep retrying | ✅                 | ✅                                 | ✅                    | Avoid excessive commit request retrying by not using the Stale Read request's start_ts to update `max_ts` |

Reviewer 2: ```suggestion
| Not use the stale read request's `start_ts` to update `max_ts` to avoid commit request keep retrying | ✅                 | ✅                                 | ❌                    | Improve commit performance in some edge cases |
```

Reviewer 2: @NingLin-P @sticnarf PTAL about whether this change is correct.

Reviewer 4: lgtm

Reviewer 5: In fact, I think it's a bug fix. If `max_ts` is updated from the `start_ts` of a stale read request, a later reader may not see committed data.

See https://github.com/pingcap/tidb/issues/25680

> | Restore many small tables would be faster.                   | ✅                 | ❌                                 | ❌                    | Split and scatter Regions concurrently to improve restore speed |

Reviewer 3: I prefer the original description, it's short and doesn't use TiDB terms 🤔

Author: Each release note is expected to start with a do-form verb. The original release note "Restore many small tables would be faster." does not accurately and adequately summarize what the PR has changed. "would" looks like a bit uncertain.

| server: stop status server early when gracefully shutdown    | ✅                 | ❌                                 | ❌                    | Shut down the status server first to make sure that the client can correctly check the shutdown status |

> | Better err msg when PD endpoint missing certificate          | ✅                 | ✅                                 | ✅                    | Improve the error message returned when a PD endpoint misses the certificate |

Reviewer 2: ```suggestion
| Better err msg when PD endpoint missing certificate          | ✅                 | ✅                                 | ✅                    | Improve the error message when connecting to a TLS protected PD endpoint without a certificate |
```

Author: ```suggestion
| Better err msg when PD endpoint missing certificate          | ✅                 | ✅                                 | ✅                    | Improve the error message when connecting to a TLS-protected PD endpoint without a certificate |
```
